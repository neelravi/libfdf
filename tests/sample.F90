#if defined HAVE_CONFIG_H
#  include "config.h"
#endif

!============================================================
!= Sample program using the f90 FDF module : September 2007 =
!============================================================
!
!     Shows FDF capabilities..
!
PROGRAM SAMPLE
  USE fdf
  USE prec
  implicit none
!--------------------------------------------------------------- Local Variables
  integer, parameter         :: maxa = 100
  logical                    :: doit, debug
  character(20)              :: fname, axis, status
  character(2)               :: symbol(maxa)
  integer(sp)                :: i, j, ia, na, external_entry
  integer(sp)                :: isa(maxa)
  real(sp)                   :: wmix
  real(dp)                   :: cutoff, phonon_energy, factor
  real(dp)                   :: xa(3, maxa)
  real(dp)                   :: listr(maxa)
  type(block_fdf)            :: bfdf
  type(parsed_line), pointer :: pline


!------------------------------------------------------------------------- BEGIN

! Initialize
  call fdf_init('sample.fdf', 'sample.out')

! Handle/Use fdf structure
  if (fdf_defined('new-style')) write(6,*) 'New-style stuff'

  na = fdf_integer('NumberOfAtoms', 0)
  write(6,*) 'Examples: na =', na

  fname = fdf_string('NameOfFile', 'whatever')
  write(6,*) 'Name of file:', fname
 
  cutoff = fdf_physical('MeshCutoff', 8.d0, 'Ry')
  write(6,*) 'MeshCutOff:', cutoff

  phonon_energy = fdf_physical('phonon-energy', 0.01d0, 'eV')
  write(6,*) 'Phonon Energy:', phonon_energy

  i = fdf_integer('SomeInt', 34)
  write(6,*) '#elems:', i

  wmix = fdf_single('WmixValue', 0.55)
  write(6,*) 'WmixValue:', wmix

  factor = fdf_double('FactorValue', 1.d-10)
  write(6,*) 'Factor:', factor

  debug = fdf_boolean('Debug', .TRUE.)
  write(6,*) 'Debug:', debug

  doit = fdf_boolean('DoIt', .FALSE.)
  write(6,*) 'Doit:', doit

  doit = fdf_defined('AtomicCoordinatesAndAtomicSpecies')
  write(6,*) 'AtomCoordsBlockDefined:', doit

  if (fdf_block('AtomicCoordinatesAndAtomicSpecies', bfdf)) then
    ia = 1
    do while((fdf_bline(bfdf, pline)) .and. (ia .le. na))
      do i= 1, 3
        xa(i,ia) = fdf_breals(pline, i)
      enddo
      isa(ia) = fdf_bintegers(pline, 1)
      ia = ia + 1
    enddo
  endif

  write(6,*) 'AtomicCoordinatesAndAtomicSpecies:'
  do ia= 1, na
    write(6,'(3F10.6,I5)') (xa(i,ia),i=1,3), isa(ia)
  enddo

  if (fdf_block('AtomicInfo', bfdf)) then
    ia = 1
    do while((fdf_bline(bfdf, pline)) .and. (ia .le. na))
      do i= 1, 3
        xa(i,ia) = fdf_breals(pline, i)
      enddo
      ia = ia + 1
    enddo
  endif

  write(6,*) 'AtomicInfo:'
  do ia= 1, na
    write(6,'(3F10.6)') (xa(i,ia),i=1,3)
  enddo

  if (fdf_block('Other-Block', bfdf)) then

!   Forward reading
    ia = 1
    do while((fdf_bline(bfdf, pline)) .and. (ia .le. na))
      symbol(ia) = fdf_bnames(pline, 1)
      do i= 1, na
        xa(i,ia) = fdf_breals(pline, i)
      enddo
      ia = ia + 1
    enddo

    write(6,*) 'Other-Block (Forward):'
    do ia= 1, na
      write(6,'(A4,3F10.6)') symbol(ia), (xa(i,ia),i=1,3)
    enddo

!   Backward reading
    ia = 1
    do while((fdf_bbackspace(bfdf, pline)) .and. (ia .le. na))
      symbol(ia) = fdf_bnames(pline, 1)
      do i= 1, na
        xa(i,ia) = fdf_breals(pline, i)
      enddo
      ia = ia + 1
    enddo

    write(6,*) 'Other-Block (Backward):'
    do ia= 1, na
      write(6,'(A4,3F10.6)') symbol(ia), (xa(i,ia),i=1,3)
    enddo

!   Forward reading
    ia = 1
    do while((fdf_bline(bfdf, pline)) .and. (ia .le. na))
      symbol(ia) = fdf_bnames(pline, 1)
      do i= 1, na
        xa(i,ia) = fdf_breals(pline, i)
      enddo
      ia = ia + 1
    enddo

    write(6,*) 'Other-Block (Forward):'
    do ia= 1, na
      write(6,'(A4,3F10.6)') symbol(ia), (xa(i,ia),i=1,3)
    enddo

!   Forward reading with rewind
    call fdf_brewind(bfdf)
    ia = 1
    do while((fdf_bline(bfdf, pline)) .and. (ia .le. na))
      symbol(ia) = fdf_bnames(pline, 1)
      do i= 1, na
        xa(i,ia) = fdf_breals(pline, i)
      enddo
      ia = ia + 1
    enddo

    write(6,*) 'Other-Block (Forward-with-rewind):'
    do ia= 1, na
      write(6,'(A4,3F10.6)') symbol(ia), (xa(i,ia),i=1,3)
    enddo
  endif

  if ( fdf_block('ListBlock',bfdf) ) then
     i = 0
     do while ( fdf_bline(bfdf,pline) )
        i = i + 1
        na = fdf_bnlists(pline)
        write(*,'(2(a,i0),a)') 'Listblock line: ',i,' has ',na,' lists'
        do ia = 1 , na
           j = -1
           call fdf_bilists(pline,ia,j,isa)
           write(*,'(tr5,2(a,i0),a)') 'list ',ia,' has ',j,' entries'
           call fdf_bilists(pline,ia,j,isa)
           write(*,'(tr5,a,1000(tr1,i0))') 'list: ',isa(1:j)
        end do
     end do
  end if

  ! Check lists
  if ( fdf_islinteger('MyList') .and. fdf_islist('MyList') &
      .and. (.not. fdf_islreal('MyList')) ) then
     na = -1
     call fdf_list('MyList',na,isa)
     if ( na < 2 ) stop 1
     write(*,'(tr1,a,i0,a)') 'MyList has ',na,' entries'
     call fdf_list('MyList',na,isa)
     write(*,'(tr5,a,1000(tr1,i0))') 'MyList: ',isa(1:na)
   else
     write(*,*)'MyList was not recognized'
     stop 1
   end if

  if ( fdf_islinteger('MyListOne') .and. fdf_islist('MyListOne') &
      .and. (.not. fdf_islreal('MyListOne')) ) then
     na = -1
     call fdf_list('MyListOne',na,isa)
     if ( na /= 1 ) stop 1
     write(*,'(tr1,a,i0,a)') 'MyListOne has ',na,' entries'
     call fdf_list('MyListOne',na,isa)
     write(*,'(tr5,a,1000(tr1,i0))') 'MyListOne: ',isa(1:na)
  else
     write(*,*)'MyListOne was not recognized'
     stop 1
  end if

  if ( fdf_islreal('MyListR') .and. fdf_islist('MyListR') &
      .and. (.not. fdf_islinteger('MyListR')) ) then
    na = -1
    call fdf_list('MyListR',na,listr)
    write(*,'(tr1,a,i0,a)') 'MyListR has ',na,' entries'
    if ( na < 2 ) stop 1
    call fdf_list('MyListR',na,listr)
    write(*,'(tr5,a,1000(tr1,f4.1))') 'MyListR: ',listr(1:na)
  else
    write(*,*)'MyListR was not recognized'
    stop 1
  end if

  if ( fdf_islreal('MyListROne') .and. fdf_islist('MyListROne') &
      .and. (.not. fdf_islinteger('MyListROne')) ) then
    na = -1
    call fdf_list('MyListROne',na,listr)
    if ( na /= 1 ) stop 1
    write(*,'(tr1,a,i0,a)') 'MyListROne has ',na,' entries'
    call fdf_list('MyListROne',na,listr)
    write(*,'(tr5,a,1000(tr1,f4.1))') 'MyListROne: ',listr(1:na)
  else
    write(*,*)'MyListROne was not recognized'
    stop 1
  end if

  if ( fdf_islist('externalentry') ) then
     write(*,*) 'externalentry is a list'
  else
     write(*,*) 'externalentry is not a list'
  end if

  external_entry = fdf_integer('externalentry', 60)
  write(6,*) 'ExternalEntry:', external_entry

  axis   = fdf_string('AxisXY', 'Cartesian')
  status = fdf_string('StatusXY', 'Enabled')
  write(6,*) 'Axis: ', TRIM(axis), ' | ', TRIM(status)

! Shutdown and deallocates fdf structure
  call fdf_shutdown()

!----------------------------------------------------------------------------END
END PROGRAM SAMPLE
