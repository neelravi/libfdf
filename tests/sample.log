 New-style stuff
 Examples: na =           3
 Name of file:Some file           
 MeshCutOff:   36.7489942245322     
 Phonon Energy:  4.264063562998147E-002
 #elems:          34
 WmixValue:  0.5500000    
 Factor:   31400000000.0000     
 Debug: F
 Doit: F
 AtomCoordsBlockDefined: T
 AtomicCoordinatesAndAtomicSpecies:
  0.000000  0.000000  0.000000    1
  1.431000  1.108000  0.000000    2
 -1.431000  1.108000  0.000000    2
 AtomicInfo:
  0.333000  0.000000  0.000000
  1.500000  1.108000  0.000000
 -1.431000  1.108000  0.000000
 Other-Block (Forward):
  Nh  1.500000  1.108000  0.000000
  Ox  0.333000  0.000000  0.000000
  Pd -1.431000  1.108000  0.000000
 Other-Block (Backward):
  Pd -1.431000  1.108000  0.000000
  Ox  0.333000  0.000000  0.000000
  Nh  1.500000  1.108000  0.000000
 Other-Block (Forward):
  Nh  1.500000  1.108000  0.000000
  Ox  0.333000  0.000000  0.000000
  Pd -1.431000  1.108000  0.000000
 Other-Block (Forward-with-rewind):
  Nh  1.500000  1.108000  0.000000
  Ox  0.333000  0.000000  0.000000
  Pd -1.431000  1.108000  0.000000
Listblock line: 1 has 3 lists
     list 1 has 5 entries
     list:  1 2 3 5 6
     list 2 has 2 entries
     list:  -1 -3
     list 3 has 13 entries
     list:  100 93 86 79 72 65 58 51 44 37 30 23 16
Listblock line: 2 has 2 lists
     list 1 has 6 entries
     list:  1 2 3 4 5 4
     list 2 has 4 entries
     list:  10 14 18 3
Listblock line: 3 has 1 lists
     list 1 has 34 entries
     list:  1 2 3 4 5 4 2 3 4 5 6 7 8 9 10 -2 -3 -4 -5 -6 -7 -8 -9 -10 -11 -12 -13 -14 -15 -16 -17 -18 -19 -20
 MyList has 11 entries
     MyList:  2 3 4 5 6 7 8 9 10 11 20
 MyListOne has 1 entries
     MyListOne:  2
 MyListR has 9 entries
     MyListR:   2.0  3.2  4.4  5.6  6.8  8.0  9.2 11.2 20.0
 MyListROne has 1 entries
     MyListROne:   2.0
 externalentry is not a list
 ExternalEntry:          99
 Axis: Polar | Enabled
PASS sample (exit status: 0)
