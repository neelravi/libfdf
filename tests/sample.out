#:defined? new-style              T
NumberOfAtoms              3
NameOfFile     Some file                                                                       
MeshCutoff          36.74899422             Ry
# above item originally: MeshCutoff          500.0000000     eV        
phonon-energy         0.4264063563E-01         eV
# above item originally: phonon-energy          344.0000000     cm**-1    
SomeInt        34     # default value
WmixValue         0.5500000119         # default value
FactorValue         0.3140000000E+11
Debug              F
DoIt              F     # default value
#:block? AtomicCoordinatesAndAtomicSpecies              T
#:defined? AtomicCoordinatesAndAtomicSpecies              T
%block AtomicCoordinatesAndAtomicSpecies
  0.000000  0.000000  0.000000  1                # Coordinates, species index
  1.431000  1.108000  0.000000  2
 -1.431000  1.108000  0.000000  2
%endblock AtomicCoordinatesAndAtomicSpecies
%block AtomicInfo
  0.333000  0.000000  0.000000
  1.500000  1.108000  0.000000
 -1.431000  1.108000  0.000000
%endblock AtomicInfo
%block Other-Block
  Nh   1.500000  1.108000  0.000000
  Ox   0.333000  0.000000  0.000000
  Pd  -1.431000  1.108000  0.000000
%endblock Other-Block
 #:(Backspace to) | Pd  -1.431000  1.108000  0.000000|
 #:(Backspace to) | Ox   0.333000  0.000000  0.000000|
 #:(Backspace to) | Nh   1.500000  1.108000  0.000000|
 #:(Backspace to) |%block Other-Block                                    # Read block|
%block Other-Block
  Nh   1.500000  1.108000  0.000000
  Ox   0.333000  0.000000  0.000000
  Pd  -1.431000  1.108000  0.000000
%endblock Other-Block
%block Other-Block
  Nh   1.500000  1.108000  0.000000
  Ox   0.333000  0.000000  0.000000
  Pd  -1.431000  1.108000  0.000000
%endblock Other-Block
%block ListBlock
  List1 [1,2,3 ,5 , 6 ] [-1,-3] [100 -- 10 step 7]
  List2 [ 1 to 5 , 4 ] [10 -- 20 step 4 , 3 ]
  List2 [ 1 to 5 , 4 , 2 to 10 , -2 to -20]
%endblock ListBlock
#:linteger? MyList              T
#:list? MyList              T
#:lreal? MyList              F
MyList             11
MyList             11
#:linteger? MyListOne              T
#:list? MyListOne              T
#:lreal? MyListOne              F
MyListOne              1
MyListOne              1
#:lreal? MyListR              T
#:list? MyListR              T
#:linteger? MyListR              F
MyListR              9
MyListR              9
#:lreal? MyListROne              T
#:list? MyListROne              T
#:linteger? MyListROne              F
MyListROne              1
MyListROne              1
#:list? externalentry              F
externalentry             99
AxisXY     Polar                                                                           
StatusXY     Enabled                                                                         
